code/core/system/proxy-velodyne16/src/velodyne16Decoder.cpp

```diff
@@ -136,14 +83,107 @@ odcore::io::conference::ContainerConference &c, const string &s, const bool &wit
+                float temp = orderedVerticalAngle[j];
+                orderedVerticalAngle[j] = orderedVerticalAngle[i];
+                orderedVerticalAngle[i] = temp;
+            }
+        }
+    }
+    //Find the sensor IDs in the odered list of vertical angles
+    for (uint8_t i = 0; i < 16; i++) {
+        for (uint8_t j = 0; j < 16; j++) {
+            if (abs(orderedVerticalAngle[i] - m_verticalAngle[j]) < 0.1f) {
+                m_sensorOrderIndex[i] = j;
+                break;
+            }
+        }
+    }      
+}
+
+void Velodyne16Decoder::setupIntensityMaskCPC(uint8_t &numberOfBitsForIntensity, uint8_t &intensityPlacement) {
+    if (numberOfBitsForIntensity != 0) {
+        m_mask = 0xFFFF;
+        if (intensityPlacement == 0) {//higher bits for intensity
+            m_mask = m_mask >> numberOfBitsForIntensity;
+        } else {
+            m_mask = m_mask << numberOfBitsForIntensity;
+        }
+    }
+}
+
+Velodyne16Decoder::Velodyne16Decoder(const std::shared_ptr< SharedMemory > m,
+odcore::io::conference::ContainerConference &c, const string &s, const bool &withCPC, const uint8_t &SPCOption, const uint8_t &CPCIntensityOption, const uint8_t &numberOfBitsForIntensity, const uint8_t &intensityPlacement, const uint8_t &distanceEncoding)
+    : m_SPCOption(SPCOption)
+    , m_CPCIntensityOption(CPCIntensityOption)
+    , m_numberOfBitsForIntensity(numberOfBitsForIntensity)
+    , m_intensityPlacement(intensityPlacement)
+    , m_mask(0)
+    , m_distanceEncoding(distanceEncoding)
+    , m_pointIndexSPC(0)
+    , m_pointIndexCPC(0)
+    , m_startID(0)
+    , m_previousAzimuth(0.0)
+    , m_currentAzimuth(0.0)
+    , m_nextAzimuth(0.0)
+    , m_deltaAzimuth(0.0)
+    , m_distance(0.0)
+    , m_velodyneSharedMemory(m)
+    , m_segment(NULL)
+    , m_velodyneContainer(c)
+    , m_spc()
+    , m_calibration(s)
+    , m_withSPC(true)
+    , m_withCPC(withCPC)
+    , m_startAzimuth(0.0)
+    , m_distanceStringStreamNoIntensity("")
+    , m_distanceStringStreamWithIntensity("")
+    , m_isStartAzimuth(true) {
+    //Initial setup of the shared point cloud (N.B. The size and width of the shared point cloud depends on the number of points of a frame, hence they are not set up in the constructor)
+    m_spc.setName(m_velodyneSharedMemory->getName()); // Name of the shared memory segment with the data.
+    m_spc.setHeight(1); // We have just a sequence of vectors.
+    m_spc.setNumberOfComponentsPerPoint(m_NUMBER_OF_COMPONENTS_PER_POINT);
+    m_spc.setComponentDataType(SharedPointCloud::FLOAT_T); // Data type per component.
+    if (m_SPCOption == 0) {
+        m_spc.setUserInfo(SharedPointCloud::XYZ_INTENSITY);
+    } else {
+        m_spc.setUserInfo(SharedPointCloud::POLAR_INTENSITY);
+    }
+
+    //Create memory for temporary storage of point cloud data for each frame
+    m_segment = (float *)malloc(m_SIZE);
+    if (m_segment == NULL) {
+        throw bad_alloc();//if memory creation using malloc fails, throw an exception
```

Comment:
The comment is not needed; my previous comment was related that the exception bad_alloc() is not helpful for the user seeing this on the console.